history/history.c

```diff
@@ -0,0 +1,1081 @@
+    sqlite3_finalize(res); res = NULL;
+error:
+    if (res)
+        sqlite3_finalize(res);
+    return rc;
+}
+
+static
+void db_free_nevra_map(struct history_nevra_map *hnm)
+{
+    if (hnm){
+        if (hnm->idmap) {
+            for (int i = 0; i < hnm->count; i++) {
+                safe_free(hnm->idmap[i]);
+            }
+            free(hnm->idmap);
+        }
+        free(hnm);
+    }
+}
+
+/*
+ * Create a map with all known NEVRAs.
+ */
+static
+struct history_nevra_map *db_nevra_map(sqlite3 *db)
+{
+    int rc = 0;
+    struct history_nevra_map *hnm = NULL;
+    sqlite3_stmt *res = NULL;
+    const char *sql_find = "SELECT * FROM rpms;";
+    int count = 0;
+
+    rc = db_rpms_maxid(db, &count);
+    check_db_rc(db, rc);
+
+    hnm = (struct history_nevra_map *)calloc(1, sizeof(struct history_nevra_map));
+    check_ptr(hnm);
+
+    hnm->idmap = (char **)calloc(count, sizeof(char *));
+    check_ptr(hnm->idmap);
+
+    hnm->count = count;
+
+    rc = sqlite3_prepare_v2(db, sql_find, -1, &res, 0);
+    check_db_rc(db, rc);
+    for(int step = sqlite3_step(res); step == SQLITE_ROW; step = sqlite3_step(res)) {
+        int id = sqlite3_column_int(res, 0);
+        char *nevra = strdup((const char *)sqlite3_column_text(res, 1));
+        check_ptr(nevra);
+        check_cond((id > 0 && id <= count));
+        /* map is zero based, rpm ids start with 1 */
+        hnm->idmap[id - 1] = nevra;
+    }
+error:
+    if (res)
+        sqlite3_finalize(res);
+    if (rc) {
+        db_free_nevra_map(hnm); hnm = NULL;
+    }
+    return hnm;
+}
+
+/*
+ * Add nevra to db if it's not there already. The id will be returned
+ * in *pid if that's not NULL.
+*/
+static
+int db_add_nevra(sqlite3 *db, const char *nevra, int *pid)
+{
+    int rc = 0, step;
+    sqlite3_stmt *res = NULL;
+    const char *sql_find = "SELECT * FROM rpms WHERE nevra = ?;";
+
+    rc = sqlite3_prepare_v2(db, sql_find, -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_text(res, 1, nevra, -1, NULL);
+    step = sqlite3_step(res);
+    if (step == SQLITE_ROW) {
+        int id = sqlite3_column_int(res, 0);
+        if (pid)
+            *pid = id;
+        sqlite3_finalize(res);
+        /* we found it, return */
+        return 0;
+    }
+    sqlite3_finalize(res); res = NULL;
+
+    /* add it to db */
+    rc = sqlite3_prepare_v2(db, "INSERT INTO rpms(nevra) VALUES (?);", -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_text(res, 1, nevra, -1, NULL);
+    step = sqlite3_step(res);
+    if (step == SQLITE_DONE) {
+        int id = sqlite3_last_insert_rowid(db);
+        if (pid)
+            *pid = id;
+    }
+
+error:
+    if (res)
+        sqlite3_finalize(res);
+    return rc;
+}
+
+/* read transaction items into ht */
+static
+int history_delta_read(sqlite3 *db, struct history_delta *hd, int id)
+{
+    int rc = 0;
+    sqlite3_stmt *res = NULL;
+    const char *sql = "SELECT * FROM trans_items WHERE trans_id = ? ORDER BY rpm_id;";
```

Comment:
use macro

Reply:
For the `SELECT` queries I disagree. It just creates more work trying to figure out what we are doing.